services/gitlab/gitlab-tag.service.js

```diff
@@ -0,0 +1,62 @@
+import Joi from 'joi'
+import { optionalUrl } from '../validators.js'
+import { NotFound } from '../index.js'
+import GitLabBase from './gitlab-base.js'
+
+const schema = Joi.array().items(
+  Joi.object({
+    name: Joi.string().required(),
+  })
+)
+
+const queryParamSchema = Joi.object({
+  gitlab_url: optionalUrl,
+}).required()
+
+export default class GitlabTag extends GitLabBase {
+  static category = 'version'
+
+  static route = {
+    base: 'gitlab/tag',
+    pattern: ':user/:repo',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'GitLab tag (latest by date)',
+      namedParams: {
+        user: 'fdroid',
+        repo: 'fdroidclient',
+      },
+      staticPreview: this.render({ name: '1.12.1' }),
+    },
+  ]
+
+  static defaultBadgeData = { label: 'tag' }
+
+  static render({ name }) {
+    return {
+      message: name,
+      color: 'blue',
+    }
+  }
+
+  async fetch({ user, repo, baseUrl }) {
+    return super.fetch({
+      schema,
+      url: `${baseUrl}/api/v4/projects/${user}%2F${repo}/repository/tags`,
+      errorMessages: {
+        404: 'repo not found',
+      },
+    })
+  }
+
+  async handle({ user, repo }, { gitlab_url: baseUrl = 'https://gitlab.com' }) {
+    const tags = await this.fetch({ user, repo, baseUrl })
+    if (tags.length === 0)
+      throw new NotFound({ prettyMessage: 'no tags found' })
+    const { name } = tags[0]
```

Comment:
This is not the latest tag by date. From the docs https://docs.gitlab.com/ee/api/tags.html

> Get a list of repository tags from a project, sorted by name in reverse alphabetical order

Not checked, but I think this is basically the same situation we've got with the GitHub API so we can maybe copy the sort by date logic from there?

Reply:
I haven't actually really looked at any part of the tag implementation, just grabbed the commits from #6389 as had been discussed in #541. Will need take a closer look

Reply:
I feel like their documentation is somewhat contradictory :thinking: The textual description of the API shared above is rather unequivocal about the sort being alphabetical by name, but then the param documentation just a couple lines below says:

> order_by | string | no | Return tags ordered by name or updated fields. Default is updated

I'm going to need to setup a test repo to try to vet it out because the handful of real repositories I was trying to spot check with were inconclusive due to both means producing the same ordering

Reply:
All signs point to the default sort indeed being by date, despite what the API description says.

https://gitlab.com/api/v4/projects/shields-ops-group%2Ftag-test/repository/tags

<details><summary>API response</summary>

```json
[
  {
    "name": "v2.0.0",
    "message": "",
    "target": "ae4abdc07357a1d26d3c692a9ad3118b6b31af2a",
    "commit": {
      "id": "ae4abdc07357a1d26d3c692a9ad3118b6b31af2a",
      "short_id": "ae4abdc0",
      "created_at": "2021-09-11T11:20:31.000-05:00",
      "parent_ids": [
        "396bf1debee4d196f85ae7f3d00ce2a4a71346a3"
      ],
      "title": "foo",
      "message": "foo\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:20:31.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:20:31.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/ae4abdc07357a1d26d3c692a9ad3118b6b31af2a"
    },
    "release": null,
    "protected": false
  },
  {
    "name": "v4.0.0",
    "message": "",
    "target": "396bf1debee4d196f85ae7f3d00ce2a4a71346a3",
    "commit": {
      "id": "396bf1debee4d196f85ae7f3d00ce2a4a71346a3",
      "short_id": "396bf1de",
      "created_at": "2021-09-11T11:16:33.000-05:00",
      "parent_ids": [
        "bf047705d78e939fcd7ff5256014c7fcee954a6f"
      ],
      "title": "another update",
      "message": "another update\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:16:33.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:16:33.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/396bf1debee4d196f85ae7f3d00ce2a4a71346a3"
    },
    "release": null,
    "protected": false
  },
  {
    "name": "v3.0.0",
    "message": "",
    "target": "bf047705d78e939fcd7ff5256014c7fcee954a6f",
    "commit": {
      "id": "bf047705d78e939fcd7ff5256014c7fcee954a6f",
      "short_id": "bf047705",
      "created_at": "2021-09-11T11:11:50.000-05:00",
      "parent_ids": [
        "490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1"
      ],
      "title": "more updates",
      "message": "more updates\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:11:50.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:11:50.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/bf047705d78e939fcd7ff5256014c7fcee954a6f"
    },
    "release": null,
    "protected": false
  },
  {
    "name": "v1.0.0",
    "message": "",
    "target": "490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1",
    "commit": {
      "id": "490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1",
      "short_id": "490f9e3d",
      "created_at": "2021-09-11T11:10:09.000-05:00",
      "parent_ids": [
        "922ab7ddafd8b7892cfa881f6d67cdde48f59ed1"
      ],
      "title": "update",
      "message": "update\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:10:09.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:10:09.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1"
    },
    "release": null,
    "protected": false
  }
]
```

</details>

Reply:
Feel like this has been sufficiently addressed in b96f597e92c1de7a9857cee5f8720c4533904382 so going to resolve, but let me know if you think there's still any outstanding items on this topic

Reply:
Odd that the docs are wrong, but I agree your assessment of the real world behaviour :+1: